File: NetFabric.CodeAnalysis.UnitTests/TestData/Enumerators.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace NetFabric.CodeAnalysis.TestData
{
    public class MissingCurrentAndMoveNextEnumerator
    {
    }

    public class MissingCurrentEnumerator
    {
        public bool MoveNext() => false;

        public ValueTask<bool> MoveNextAsync() => new ValueTask<bool>(Task.FromResult(false));
    }

    public class MissingMoveNextEnumerator<T>
    {
        public T Current => default;
    }

    public class Enumerator<T>
    {
        public T Current => default;

        public bool MoveNext() => false;
    }

    public class AsyncEnumerator<T>
    {
        public T Current => default;

        public ValueTask<bool> MoveNextAsync() => new ValueTask<bool>(Task.FromResult(false));
    }

    public class ExplicitEnumerator<T> : IEnumerator<T>
    {
        T IEnumerator<T>.Current => default;
        object IEnumerator.Current => default;

        bool IEnumerator.MoveNext() => false;

        void IEnumerator.Reset() { }

        void IDisposable.Dispose() { }
    }

    public class ExplicitAsyncEnumerator<T> : IAsyncEnumerator<T>
    {
        T IAsyncEnumerator<T>.Current => default;

        ValueTask<bool> IAsyncEnumerator<T>.MoveNextAsync() => new ValueTask<bool>(Task.FromResult(false));

        ValueTask IAsyncDisposable.DisposeAsync() => new ValueTask();
    }
}

File: NetFabric.VisualBasic.UnitTests/TestData/Enumerators.vb
Namespace TestData

    Public Structure MappedEnumerator(Of T)
        Implements IEnumerator(Of T)

        Private ReadOnly _source As T()
        Private _index As Integer

        Public Sub New(source As T())
            _source = source
            _index = -1
        End Sub

        Public ReadOnly Property Current As T Implements IEnumerator(Of T).Current
            Get
                Current = _source(_index)
            End Get
        End Property

        Public ReadOnly Property IEnumerator_Current As Object Implements IEnumerator.Current
            Get
                IEnumerator_Current = _source(_index)
            End Get
        End Property

        Public Function MoveNext() As Boolean Implements IEnumerator.MoveNext
            _index = _index + 1
            MoveNext = _index < _source.Length
        End Function

        Public Sub Reset() Implements IEnumerator.Reset
            Throw New NotSupportedException
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
        End Sub

    End Structure

End Namespace

